Fix #3015 : In the simple editor, add the ability to highlight unfinished questions.

SudhanvaMG opened the pull request:
SidebarHighlightingService created ! . It has two functions which will change the required sidebar element to error mode and vice versa .

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/sidebarHighlightingService.js:
```diff
@@ -0,0 +1,46 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Service that highlights the color of the question
```
We want this to be used in other cases too, so 'Service for managing sidebar items highlighting.' or similar.

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/sidebarHighlightingService.js:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * @fileoverview Service that highlights the color of the question
+ * depending on wheather it is finished or unfinished .
```
This is somehow misleading since it only provides functionality to do the highlighting.

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/sidebarHighlightingService.js:
```diff
@@ -0,0 +1,46 @@
+    var ERROR_CLASS = 'error';
+
+    return {
+      errorHighlight: function(id) {
```
Please add a comment to explain usage of each method.

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/sidebarHighlightingService.js:
```diff
@@ -0,0 +1,46 @@
+            return;
+          } else {
+            var elm = angular.element(document.getElementById(
+              QuestionIdService.SIDEBAR_PREFIX + id)
```
You forgot to inject the QuestionIdService.

seanlip replied:
@SudhanvaMG -- how did this slip by your manual testing?

SudhanvaMG replied:
@seanlip I dont know . I didnt get any errors regarding this .

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/sidebarHighlightingService.js:
```diff
@@ -0,0 +1,46 @@
+            elm.addClass(ERROR_CLASS);
+          }
+        },
+      undoErrorHighlight: function(id) {
```
This method takes and id, but it does not use it, it removes the 'error' class from all sidebar items that have it, i think when using this function we will probably want to undo only one.

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/simple_editor_sidebar_directive.html:
```diff
@@ -112,6 +112,9 @@
   simple-editor-sidebar .highlighted {
     background-color: #c5e1a5;
   }
+  simple-editor-sidebar .error {
```
Error highlighting is not done by changing the background, see [mocks ](https://drive.google.com/drive/folders/0B_8dp8q8tZvqSWtqRkpTY0djckU) please.

SudhanvaMG replied:
@andromfins the text color will change. Need to add the icon also ?

himanshu-dixit replied:
Yes sudhavanaMG, icon is needed to visually indicate about the error.

SudhanvaMG replied:
@himanshu-dixit  Ok will add the icon .

himanshu-dixit replied:
@SudhanvaMG Probably we're not using font-awesome (probably not 100% sure). Most icons at this point are thru SVG or Unicode symbol. Check this out, http://www.fileformat.info/info/unicode/char/26a0/index.htm

andromfins replied:
@SudhanvaMG You can use [material icons](https://material.io/icons/#ic_error_outline).

SudhanvaMG replied:
@andromfins Is material icons already set up and been used ?

seanlip replied:
@SudhanvaMG: have you tried grepping?

SudhanvaMG replied:
@andromfins  Adding the icon doesnt come in the class . We will have to add a new <i> tag which will show only when an error comes up . So should i do that in this issue or  a different one ?

seanlip replied:
Then please do so. That is the first thing one should do when trying to figure out whether something exists in the code or not, or where some functionality is. Only if you can't find it this way, should you ask (and also explain what you've tried).

In case it helps, you can use a bash invocation similar to: `grep "thing to search for" . -r --exclude-dir=third_party`. The 'thing to search for' here could be 'material', for example.

Re your question to @andromfins, I think that this PR should match the error-handling part of the mocks and should implement whatever is needed to do so, including any necessary icons.

SudhanvaMG replied:
@seanlip Yes i have searched the code whenever required but i just didnt use grepping , my editor has directory and file searches which i have been using from start to figure where a function call has been made and how different files are connected.

seanlip replied:
Ah, OK, sorry -- I misunderstood you. Could you please search for "material"?

SudhanvaMG replied:
Yes , i did and found out where and how material icons have been used. 
The thing is if i implement the icon feature in this issue , i need to work on finding out when the error status come up so that a scope variable $scope.isError is False and when error occurs it is set to true and using ng-show we can make to icon visible for the question with error .
So this will belong to a different issue as we need to write function which checks for the error .

seanlip replied:
OK, glad to hear that you figured out how to use the material icons!

I'll defer to @andromfins on your other question, as I don't think I have enough context to answer it, apart from the general principle I mentioned a couple of comments before. But the fact that you would need to write additional functions in order to implement the full functionality doesn't affect that comment.

SudhanvaMG replied:
@seanlip Ok i got it . I will talk with @andromfins about this .
Thank you !

seanlip commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/ScrollSyncService.js:
```diff
@@ -101,4 +101,4 @@ oppia.factory('ScrollSyncService', [
 
     return scrollSyncService;
-}]);
+  }]);
```
Ah. Break after '}' so that opening and closing brackets match.

SudhanvaMG replied:
Its giving an error if i break after '{'

seanlip replied:
What is the error?

seanlip replied:
(I just made this change, and there was no error.)

seanlip commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SimpleEditorSidebarDirective.js:
```diff
@@ -25,30 +25,30 @@ oppia.directive('simpleEditorSidebar', [function() {
-        $scope.$apply();
-      });
-    }]
+          $scope.SUBFIELD_LABELS = [
```
@andromfins -- can you (or someone else) please take care of all these lint issues in a separate PR that's dedicated just to doing that? I think @himanshu-dixit also commented on them, and we really shouldn't leave this branch in an unlinted state to begin with.

seanlip commented:
That probably means that this "execution branch" of your code was not run
when you were doing manual testing. Can you make sure it is? You can add
console logs (temporarily) to ensure that your manual testing covers all
code paths.

On Wednesday, February 15, 2017, Sudhanva MG <notifications@github.com>
wrote:

> *@SudhanvaMG* commented on this pull request.
> ------------------------------
>
> In core/templates/dev/head/pages/exploration_editor/simple_editor_tab/
> sidebarHighlightingService.js <https://github.com/oppia/oppia/pull/3058>:
>
> > + * @fileoverview Service that highlights the color of the question
> + * depending on wheather it is finished or unfinished .
> + */
> +
> +oppia.factory('SidebarHighlightingService', [
> +  function() {
> +    var ERROR_SELECTOR = 'simple-editor-sidebar .error';
> +    var ERROR_CLASS = 'error';
> +
> +    return {
> +      errorHighlight: function(id) {
> +          if (!angular.isDefined(id)) {
> +            return;
> +          } else {
> +            var elm = angular.element(document.getElementById(
> +              QuestionIdService.SIDEBAR_PREFIX + id)
>
> @seanlip <https://github.com/seanlip> I dont know . I didnt get any
> errors regarding this .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3058>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyuxu9DauumKo0YLT0if2wWkD2B_1ks5rcuktgaJpZM4MAOeg>
> .
>

SudhanvaMG commented:
Yes, i checked it using browser console.

On Thu, Feb 16, 2017 at 12:38 AM, Sean Lip <notifications@github.com> wrote:

> That probably means that this "execution branch" of your code was not run
> when you were doing manual testing. Can you make sure it is? You can add
> console logs (temporarily) to ensure that your manual testing covers all
> code paths.
>
> On Wednesday, February 15, 2017, Sudhanva MG <notifications@github.com>
> wrote:
>
> > *@SudhanvaMG* commented on this pull request.
> > ------------------------------
> >
> > In core/templates/dev/head/pages/exploration_editor/simple_editor_tab/
> > sidebarHighlightingService.js <https://github.com/oppia/oppia/pull/3058
> >:
> >
> > > + * @fileoverview Service that highlights the color of the question
> > + * depending on wheather it is finished or unfinished .
> > + */
> > +
> > +oppia.factory('SidebarHighlightingService', [
> > + function() {
> > + var ERROR_SELECTOR = 'simple-editor-sidebar .error';
> > + var ERROR_CLASS = 'error';
> > +
> > + return {
> > + errorHighlight: function(id) {
> > + if (!angular.isDefined(id)) {
> > + return;
> > + } else {
> > + var elm = angular.element(document.getElementById(
> > + QuestionIdService.SIDEBAR_PREFIX + id)
> >
> > @seanlip <https://github.com/seanlip> I dont know . I didnt get any
> > errors regarding this .
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/oppia/oppia/pull/3058>, or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AKFeyuxu9DauumKo0YLT0if2wWkD2B_1ks5rcuktgaJpZM4MAOeg>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3058#issuecomment-280107092>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARDdE4JVod2F8MHfB8wfTIUOTWIiwvmXks5rc00pgaJpZM4MAOeg>
> .
>

SudhanvaMG commented:
This is how it looks now without the icon !
![screenshot from 2017-02-16 14-10-08](https://cloud.githubusercontent.com/assets/17882387/23013678/b47526a4-f451-11e6-98ea-c5297cab4ede.png)

himanshu-dixit commented:
@SudhanvaMG The font color seems to bit different, according to mocks #f82d4e needs to be used. This color tends to bit sharp, also for font awesome icon you can refer to third_party.css .

andromfins commented:
@SudhanvaMG Please do some manual testing, something like:
  - Create a new exploration, fill up the title, introduction, add a couple questions.
  - open up the browser console.
  - Call the SidebarHighlightingService methods to errorHighlight a highlighted (green) sidebar item and a non highlighted sidebaritem.
  - Undo the highlighting of these elements. e.g:
  ```Javascript
  $injector = angular.element('simple-editor-body').injector();
  sHS = $injector.get('SidebarHighlightingService');
  sHS.errorHighlight('sidebaritem:Introduction:correctanswer');
  sHS.undoErrorHighlight('sidebaritem:Introduction:correctanswer');
  ```

- Currently the color property of the .error class is overwritten, undoErrorHighlighting raises an error (see screenshot), and the icon need to be positioned.
![review-3058](https://cloud.githubusercontent.com/assets/10903202/23073356/19dbc8a0-f52d-11e6-9c72-f1f82a98c73a.png)

SudhanvaMG commented:
@andromfins  I have fixed the color issue . Please review it .

andromfins commented:
@SudhanvaMG Here are some things that i noticed.

 - The icon is not on the same line as the text (vertical position).
 - It would be preferable that the horizontal position of the icon, to not be entirely fixed for all cases (In case of Hints for example it looks a bit too far) so if it could just have a certain fixed left padding it would be better.
 - The service should be able to highlight Title and Introduction too (the color problem persists for these two elements).
 - Some icons look red, please preserve the overall existing look while making changes.
 - undoErrorHighlight don't work as expected
```Javascript
$injector = angular.element('simple-editor-body').injector();
sHS = $injector.get('SidebarHighlightingService');
sHS.errorHighlight('sidebaritem:Introduction:correctanswer');
sHS.errorHighlight('sidebaritem:Introduction:hints');
sHS.undoErrorHighlight('sidebaritem:Introduction:hints');
```
The code above should leave us with the Correct Answer element on Question 1 highlighted, which is not the case.
![review-3058-2](https://cloud.githubusercontent.com/assets/10903202/23175755/348338b8-f858-11e6-9cab-7b72210b4f98.png)

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SidebarHighlightingService.js:
```diff
@@ -0,0 +1,50 @@
+          } else {
+            var elm = angular.element(document.getElementById(id));
+            elm.removeClass(ERROR_CLASS);
+            $('#icon_elm').remove();
```
This is the reason undoErrorHighlight deletes the icon from the non desirable element, just remove the second child of elm.

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/simple_editor_sidebar_directive.html:
```diff
@@ -134,4 +140,10 @@
     text-indent: 30px;
     width: 220px;
   }
+  .material-icons.md-18 {
```
Here you override [this](https://github.com/oppia/oppia/blob/simple-editor/core/templates/dev/head/css/oppia.css#L71), which causes all icons with this class to change the color, use something like 'simple-editor-sidebar .error-outline-icon'

andromfins commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SidebarHighlightingService.js:
```diff
@@ -0,0 +1,50 @@
+  'QuestionIdService',
+  function(QuestionIdService) {
+    var ERROR_CLASS = 'error';
+    var ICON_ELM = '<i id="icon_elm" class="material-icons md-18">&#xE001;</i>';
```
Please change the class here to `class="material-icons error-outline-icon"` and use it bellow as explained.

SudhanvaMG commented:
@andromfins  Ok will look into it and update the code .

SudhanvaMG commented:
Hey @andromfins I have fixed all the above things you have mentioned 

- Aligned the icon as the same line as text .

- The dash color changed to red.

- The left padding for the icon cannot be fixed because , as i am appending the icon element to the li tag , the icon gets appended to the next line so if have used top and left positions to align the icon on the same line as text (horizontally ) and same line as all other icons (vertically) .

- The service is working for title and intro also .

- I have changed the class of the icon so that it doesnt overlap with other md-icons.

- undoErrorHighligh is working properly .

![screenshot from 2017-02-24 10-04-40](https://cloud.githubusercontent.com/assets/17882387/23291668/f7824f6a-fa7f-11e6-9e30-e311a77e7ec1.png)

![screenshot from 2017-02-24 10-49-10](https://cloud.githubusercontent.com/assets/17882387/23291679/0f14be56-fa80-11e6-89df-56c5f2ec4c0a.png)

andromfins commented:
Looks great, thanks !

seanlip commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SidebarHighlightingService.js:
```diff
@@ -0,0 +1,57 @@
+            var elm = angular.element(document.getElementById(id));
+            // Append the id to the icon so that we can uniquely identify
+            // the icon tag.
+            var ICON_ELM = '<i id="icon_elm_' + id +
```
OK -- this needs to be changed. I'm sorry I did not spot this before it was merged, but we cannot be doing HTML construction in the JS. It's unsafe. If for some reason the id string ever had a ">" character then we would be running into possible XSS issues.

I'll see if I can do a follow-up commit to address this, but please bear it in mind for the future.

SudhanvaMG replied:
@seanlip  I had already mentioned the alternate way in the above comment , but as that will belong to a different issue , we had to use this .

seanlip commented on core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SidebarHighlightingService.js:
```diff
@@ -0,0 +1,57 @@
+    return {
+      // Highlights the sidebar item with error status
+      errorHighlight: function(id) {
+          if (!angular.isDefined(id)) {
```
The indentation here is wrong.

seanlip commented:
This PR needs to be reverted. There are a number of serious issues in the code and the newly-added service doesn't appear to be being used. Nothing is calling errorHighlight.

@andromfins -- did you actually test this before merging? This is the sort of thing that needs to be checked for, during review.

seanlip commented:
Update: I have reverted this PR.

SudhanvaMG commented:
@seanlip Yes , we havent called errorhighligt anywhere , this issue was only to address the function that will change the color and add the icon . This function will be called elsewhere when we are checking for errors  when the publish button is pressed .
